bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/PacketListener.java

```diff
@@ -0,0 +1,130 @@
+    private final List<PayloadHandler> handlers = new CopyOnWriteArrayList<>();
+
+    private String multicastGroup;
+    private int port;
+
+    public static final String DEFAULT_MCAST_GRP = "239.12.255.254";
+    public static final int DEFAULT_MCAST_PORT = 9522;
+
+    private MulticastSocket socket;
+    private ScheduledFuture<?> future;
+
+    public PacketListener(DefaultPacketListenerRegistry registry, String multicastGroup, int port) {
+        this.registry = registry;
+        this.multicastGroup = multicastGroup;
+        this.port = port;
+    }
+
+    public void addPayloadHandler(PayloadHandler handler) {
+        handlers.add(handler);
+    }
+
+    public void removePayloadHandler(PayloadHandler handler) {
+        handlers.remove(handler);
+
+        if (handlers.isEmpty()) {
+            registry.close(multicastGroup, port);
+        }
+    }
+
+    public boolean isOpen() {
+        return socket != null && socket.isConnected();
+    }
+
+    public void open(int intervalSec) throws IOException {
+        if (isOpen()) {
+            // no need to bind socket second time
+            return;
+        }
+        socket = new MulticastSocket(port);
+        socket.setSoTimeout(5000);
+        InetAddress address = InetAddress.getByName(multicastGroup);
+        socket.joinGroup(address);
+
+        future = registry.addTask(new ReceivingTask(socket, multicastGroup + ":" + port, handlers), intervalSec);
+    }
+
+    void close() throws IOException {
+        if (future != null) {
+            future.cancel(true);
+        }
+
+        InetAddress address = InetAddress.getByName(multicastGroup);
+        socket.leaveGroup(address);
+        socket.close();
+    }
+
+    public void request() {
+        registry.execute(new ReceivingTask(socket, multicastGroup + ":" + port, handlers));
+    }
+
+    static class ReceivingTask implements Runnable {
+        private final Logger logger = LoggerFactory.getLogger(ReceivingTask.class);
+        private final DatagramSocket socket;
+        private final String group;
+        private final List<PayloadHandler> handlers;
+
+        ReceivingTask(DatagramSocket socket, String group, List<PayloadHandler> handlers) {
+            this.socket = socket;
+            this.group = group;
+            this.handlers = handlers;
+        }
+
+        public void run() {
+            try {
+                byte[] bytes = new byte[608];
+                DatagramPacket msgPacket = new DatagramPacket(bytes, bytes.length);
+                socket.receive(msgPacket);
+
+                try {
+                    EnergyMeter meter = new EnergyMeter();
+                    meter.parse(bytes);
+
+                    for (PayloadHandler handler : handlers) {
+                        handler.handle(meter);
+                    }
+                } catch (IOException e) {
+                    logger.info("Unexpected payload received for group {}", group, e);
```

Comment:
Should be lowered to debug or trace level!

Reply:
Indeed.